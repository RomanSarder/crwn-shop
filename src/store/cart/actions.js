import { createAction } from "@reduxjs/toolkit";

export var toggleCart = createAction('cart/toggle')
export var addItemToCart = createAction('cart/addItem')
export var removeItemFromCart = createAction('cart/removeItem')
export var increaseItemQuantity = createAction('cart/increaseQuantity')
export var decreaseItemQuantity = createAction('cart/decreaseQuantity')