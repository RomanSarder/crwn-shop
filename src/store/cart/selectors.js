export function selectCartDisplayState (state) {
    return state.cart.showCart
}