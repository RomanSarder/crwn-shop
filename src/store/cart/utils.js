export function getCartItemById (state, id) {
    const itemIndex = state.cartItems.findIndex(function findPendingItem (item) {
        return item.id == id
    })

    var targetItem = state.cartItems[itemIndex]

    return {
        index: itemIndex,
        item: targetItem
    }
}

export function removeCartItem (state, id) {
    return state.cartItems.filter(function filterInItemsExceptId (item) {
        return item.id !== id
    })
}