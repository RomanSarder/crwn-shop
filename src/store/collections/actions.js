import { createAction, createAsyncThunk } from "@reduxjs/toolkit";
import { getCollectionsWithItems } from "../../api/collections";

export const setCollectionsData = createAction('collections/setData')
export const collectionLoadingFinished = createAction('collections/loadingFinished')
export const collectionLoadingStarted = createAction('collections/loadingStarted')

export const getCollectionsData = createAsyncThunk(
    'collections/getData',
    async function getData (_, { dispatch }) {
        dispatch(collectionLoadingStarted())
        
        try {
            var response = await getCollectionsWithItems()
            dispatch(setCollectionsData(response))
        } finally {
            dispatch(collectionLoadingFinished())
        }
    },
    {
        condition: function checkIfDataAlreadyLoaded (_, { getState }) {
          const { collections: { data, isFetching } } = getState()
          const isDataAlreadyLoaded = Object.keys(data).length > 0
          if (isDataAlreadyLoaded || isFetching) {
            // Already in progress or data was loaded
            return false
          }
        }
    }
)