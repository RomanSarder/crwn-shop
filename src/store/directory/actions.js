import { createAsyncThunk, createAction } from "@reduxjs/toolkit";
import { getDirectoryData as getDirectoryDataFromAPI } from "../../api/directory";

export var setDirectoryData = createAction('directory/setData')
export var directoryLoadingFinished = createAction('directory/loadingFinished')
export var directoryLoadingStarted = createAction('directory/loadingStarted')

export var getDirectoryData = createAsyncThunk(
    'directory/getData',
    async function getData (_, { dispatch }) {
        dispatch(directoryLoadingStarted())
        
        try {
            var response = await getDirectoryDataFromAPI()
            dispatch(setDirectoryData(response))
        } finally {
            dispatch(directoryLoadingFinished())
        }
    },
    {
        condition: function checkIfDataAlreadyLoaded (_, { getState }) {
          var { directory: { items, isFetching } } = getState()
          const isDataAlreadyLoaded = items.length > 0
          if (isDataAlreadyLoaded || isFetching) {
            // Already in progress or data was loaded
            return false
          }
        }
    }
)