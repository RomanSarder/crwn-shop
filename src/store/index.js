import { configureStore } from '@reduxjs/toolkit'
import logger from 'redux-logger'

import userReducer from './user/userSlice'

export default configureStore({
    reducer: {
        user: userReducer
    },
    middleware: [logger]
})