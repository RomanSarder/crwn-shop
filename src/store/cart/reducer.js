import { createReducer } from "@reduxjs/toolkit";
import { 
        addItemToCart as addItemToCartAction, 
        decreaseItemQuantity as decreaseItemQuantityAction, 
        increaseItemQuantity as increaseItemQuantityAction, 
        removeItemFromCart as removeItemFromCartAction, 
        toggleCart as toggleCartAction 
        } from "./actions";
import { getCartItemById, removeCartItem } from "./utils";

var initialState = {
    showCart: false,
    cartItems: [],
}

export default createReducer(initialState, function buildReducer (builder) {
    builder
        .addCase(toggleCartAction, function updateState (state) {
            state.showCart = !state.showCart
        })
        .addCase(addItemToCartAction, function updateState (state, { payload }) {
            var pendingItem = { ...payload, quantity: 1 }

            var { index: targetItemInCartIndex, item: targetItemInCart } = getCartItemById(state, pendingItem.id)
            const isItemAlreadyInCart = targetItemInCartIndex > -1
            if (isItemAlreadyInCart) {
                state.cartItems[targetItemInCartIndex] = { ...targetItemInCart, quantity: targetItemInCart.quantity + 1 }
            } else {
                state.cartItems.push({ ...pendingItem, quantity: 1 })
            }
        })
        .addCase(removeItemFromCartAction, function updateState (state, { payload: { id } }) {
            state.cartItems = removeCartItem(state, id)
        })
        .addCase(increaseItemQuantityAction, function updateState (state, { payload: { id } }) {
            var { index, item } = getCartItemById(state, id)
            state.cartItems[index].quantity = item.quantity + 1
        })
        .addCase(decreaseItemQuantityAction, function updateState (state, { payload: { id } }) {
            var { index, item } = getCartItemById(state, id)

            if (item.quantity == 1) {
                state.cartItems = removeCartItem(state, id)
            } else {
                state.cartItems[index].quantity = item.quantity - 1
            }
        })
})