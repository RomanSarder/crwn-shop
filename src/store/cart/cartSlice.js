import { persistReducer } from 'redux-persist'
import { cartPersistConfig } from '../persist-config'
import reducer from './reducer'

export * as actions from './actions'
export * as selectors from './selectors'
export default persistReducer(cartPersistConfig, reducer)