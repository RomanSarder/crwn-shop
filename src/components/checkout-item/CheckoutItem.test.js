import React from 'react'
import { render } from '@testing-library/react'
import 'jest-styled-components'
import CheckoutItemContainer, { CheckoutItem } from './CheckoutItem'
import userEvent from '@testing-library/user-event'
import { createStoreInstance } from '../../store'
import { addItemToCart } from '../../store/cart/actions'
import { Provider } from 'react-redux'

var testCheckoutItem = {
    id: 2,
    name: 'test item',
    quantity: 2,
    price: 5,
    imageUrl: 'test image url'
}

function renderCheckoutItemComponent ({ remove = jest.fn(), decreaseQuantity = jest.fn(), increaseQuantity = jest.fn(), item = testCheckoutItem } = {}) {
    var renderResult = render(
        <CheckoutItem item={item} remove={remove} decreaseQuantity={decreaseQuantity} increaseQuantity={increaseQuantity} />
    )

    return renderResult
}

function renderCheckoutItemContainer (store) {
    var renderResult = render(
        <Provider store={store}>
            <CheckoutItemContainer item={testCheckoutItem} />
        </Provider>
    )

    return renderResult
}

it('should render checkout item properly', () => {
    var { container, getByAltText, getByText } = renderCheckoutItemComponent()

    expect(container.firstChild).toMatchSnapshot()
    expect(getByAltText(testCheckoutItem.name)).toBeVisible()
    expect(getByAltText(testCheckoutItem.name)).toHaveAttribute('src', testCheckoutItem.imageUrl)
    expect(getByText(testCheckoutItem.name)).toBeVisible()
    expect(getByText(testCheckoutItem.quantity)).toBeVisible()
    expect(getByText(testCheckoutItem.price)).toBeVisible()
})

it('should have working buttons', () => {
    var removeMock = jest.fn()
    var decreaseQuantityMock = jest.fn()
    var increaseQuantityMock = jest.fn()

    var { getByText } = renderCheckoutItemComponent({ 
        decreaseQuantity: decreaseQuantityMock,
        increaseQuantity: increaseQuantityMock,
        remove: removeMock
    })

    userEvent.click(getByText('❮'))
    expect(decreaseQuantityMock).toHaveBeenCalledTimes(1)
    userEvent.click(getByText('❯'))
    expect(increaseQuantityMock).toHaveBeenCalledTimes(1)
    userEvent.click(getByText('✕'))
    expect(removeMock).toHaveBeenCalledTimes(1)
})

it('should properly render with redux state', () => {
    var store = createStoreInstance()

    store.dispatch(addItemToCart(testCheckoutItem))
    var { getByAltText, getByText, container } = renderCheckoutItemContainer(store)

    expect(container.firstChild).toMatchSnapshot()
    expect(getByAltText(testCheckoutItem.name)).toBeVisible()
    expect(getByAltText(testCheckoutItem.name)).toHaveAttribute('src', testCheckoutItem.imageUrl)
    expect(getByText(testCheckoutItem.name)).toBeVisible()
    expect(getByText(testCheckoutItem.quantity)).toBeVisible()
    expect(getByText(testCheckoutItem.price)).toBeVisible()    
})

it('should properly decrease item quantity', () => {
    var store = createStoreInstance()
    store.dispatch(addItemToCart(testCheckoutItem))
    store.dispatch(addItemToCart(testCheckoutItem))
    var { getByText } = renderCheckoutItemContainer(store)
    userEvent.click(getByText('❮'))
    expect(store.getState().cart.cartItems[0].quantity).toEqual(1)
})

it('should properly increase item quantity', () => {
    var store = createStoreInstance()
    store.dispatch(addItemToCart(testCheckoutItem))
    var { getByText } = renderCheckoutItemContainer(store)
    userEvent.click(getByText('❯'))
    expect(store.getState().cart.cartItems[0].quantity).toEqual(2)
})

it('should properly remove cart item', () => {
    var store = createStoreInstance()
    store.dispatch(addItemToCart(testCheckoutItem))
    var { getByText } = renderCheckoutItemContainer(store)
    userEvent.click(getByText('✕'))
    expect(store.getState().cart.cartItems.length).toEqual(0)
})