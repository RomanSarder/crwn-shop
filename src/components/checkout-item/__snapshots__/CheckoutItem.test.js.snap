// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should properly render with redux state 1`] = `
.c2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  background: none;
  border: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c2:focus {
  outline: none;
}

.c0 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  min-height: 10rem;
  border-bottom: 1px solid darkgrey;
  padding: 1.5rem 0;
  font-size: 2rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 .image-container {
  width: 23%;
  padding-right: 1.5rem;
}

.c0 .image-container img {
  width: 100%;
  max-height: 100%;
}

.c0 .name,
.c0 .quantity,
.c0 .price {
  width: 23%;
  text-align: center;
}

.c0 .quantity {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 .quantity .c1 {
  font-size: 1.6rem;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c0 .remove-button {
  padding-left: 1.2rem;
  cursor: pointer;
}

.c0 .c1 {
  font-size: 2.6rem;
}

<div
  class="c0"
>
  <div
    class="image-container"
  >
    <img
      alt="test item"
      src="test image url"
    />
  </div>
  <span
    class="name"
  >
    test item
  </span>
  <div
    class="quantity"
  >
    <button
      class="c1 c2"
    >
      <span>
        ❮
      </span>
    </button>
    <span
      class="value"
    >
      2
    </span>
    <button
      class="c1 c2"
    >
      <span>
        ❯
      </span>
    </button>
  </div>
  <span
    class="price"
  >
    5
  </span>
  <div
    class="remove-button"
  >
    <button
      class="c1 c2"
    >
      ✕
    </button>
  </div>
</div>
`;

exports[`should render checkout item properly 1`] = `
.c2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  background: none;
  border: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c2:focus {
  outline: none;
}

.c0 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  min-height: 10rem;
  border-bottom: 1px solid darkgrey;
  padding: 1.5rem 0;
  font-size: 2rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 .image-container {
  width: 23%;
  padding-right: 1.5rem;
}

.c0 .image-container img {
  width: 100%;
  max-height: 100%;
}

.c0 .name,
.c0 .quantity,
.c0 .price {
  width: 23%;
  text-align: center;
}

.c0 .quantity {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 .quantity .c1 {
  font-size: 1.6rem;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c0 .remove-button {
  padding-left: 1.2rem;
  cursor: pointer;
}

.c0 .c1 {
  font-size: 2.6rem;
}

<div
  class="c0"
>
  <div
    class="image-container"
  >
    <img
      alt="test item"
      src="test image url"
    />
  </div>
  <span
    class="name"
  >
    test item
  </span>
  <div
    class="quantity"
  >
    <button
      class="c1 c2"
    >
      <span>
        ❮
      </span>
    </button>
    <span
      class="value"
    >
      2
    </span>
    <button
      class="c1 c2"
    >
      <span>
        ❯
      </span>
    </button>
  </div>
  <span
    class="price"
  >
    5
  </span>
  <div
    class="remove-button"
  >
    <button
      class="c1 c2"
    >
      ✕
    </button>
  </div>
</div>
`;
